kernel/kernel-api/src/main/java/io/delta/kernel/internal/IcebergCompatV2Utils.java

```diff
@@ -33,4 +66,216 @@ public static void validateDataFileStatus(DataFileStatus dataFileStatus) {
+  private static class RequiredDeltaTablePropertyEnforcer {
+    public final String propertyName;
+    public final Predicate<String> validator;
+    public final String autoSetValue;
+    public final PostMetadataProcessor postMetadataProcessor;
+
+    /**
+     * Constructor for RequiredDeltaTableProperty
+     *
+     * @param propertyName DeltaConfig we are checking
+     * @param validator A generic method to validate the given value
+     * @param autoSetValue The value to set if we can auto-set this value (e.g. during table
+     *     creation)
+     * @param postMetadataProcessor A callback to post-process the metadata
+     */
+    RequiredDeltaTablePropertyEnforcer(
+        String propertyName,
+        Predicate<String> validator,
+        String autoSetValue,
+        PostMetadataProcessor postMetadataProcessor) {
+      this.propertyName = propertyName;
+      this.validator = validator;
+      this.autoSetValue = autoSetValue;
+      this.postMetadataProcessor = postMetadataProcessor;
+    }
+
+    Optional<Metadata> validateAndUpdate(boolean isCreatingTable, Metadata newMetadata) {
+      String newestValue = newMetadata.getConfiguration().get(propertyName);
+      boolean newestValueOkay = validator.test(newestValue);
+      boolean newestValueExplicitlySet = newMetadata.getConfiguration().containsKey(propertyName);
+
+      if (!newestValueOkay) {
+        if (!newestValueExplicitlySet && isCreatingTable) {
+          // Covers the case CREATE that did not explicitly specify the required table property.
+          // In these cases, we set the property automatically.
+          newMetadata = newMetadata.withNewConfiguration(singletonMap(propertyName, autoSetValue));
+          return Optional.of(newMetadata);
+        } else {
+          // In all other cases, if the property value is not compatible
+          // with the IcebergV1 requirements, we fail
+          throw new UnsupportedOperationException(
+              String.format(
+                  "The value '%s' for the property '%s' is not compatible with "
+                      + "IcebergV2 requirements",
+                  newestValue, propertyName));
+        }
+      }
+
+      return Optional.empty();
+    }
+  }
+
+  private abstract static class IcebergCompatCheck {
+    abstract void check(boolean isCreatingNewTable, Metadata newMetadata, Protocol newProtocol);
+  }
+
+  private static final RequiredDeltaTablePropertyEnforcer ICEBERG_COMPAT_V2_CM_REQUIREMENT =
+      new RequiredDeltaTablePropertyEnforcer(
+          TableConfig.COLUMN_MAPPING_MODE.getKey(),
+          (value) ->
+              value.equals(ColumnMappingMode.NAME.name())
+                  || value.equals(ColumnMappingMode.ID.name()),
+          ColumnMappingMode.NAME.value,
+          new PostMetadataProcessor() {
+            @Override
+            Optional<Metadata> postProcess(Metadata newMetadata, boolean isCreatingNewTable) {
+              // TODO: assign column mapping field ids
```

Comment:
Wouldn't we do this elsewhere later on if we just enable column mapping mode in the metadata? (We have to do this for non-IcebergCompatV2 tables anyways right)

Reply:
Yes, we assign the field IDs if they are _not_ already assigned. Calling it multiple times is not an issue. Also this stems from the lack of specific order of checking and making metadata updates.